cpp/src/arrow/engine/substrait/relation_internal.cc

```diff
@@ -338,15 +409,29 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
       if (!left_keys || !right_keys) {
         return Status::Invalid("Left keys for join cannot be null");
       }
+
+      // Create output schema from left, right relations and join keys
+      std::shared_ptr<Schema> join_schema = left.output_schema;
+      std::shared_ptr<Schema> right_schema = right.output_schema;
+
+      for (const auto& field : right_schema->fields()) {
+        ARROW_ASSIGN_OR_RAISE(
+            join_schema, join_schema->AddField(
+                             static_cast<int>(join_schema->fields().size()) - 1, field));
+      }
```

Comment:
Each time you call `AddField` it is going to create a new schema, create a new vector of fields, and copy all the fields over.  So this ends up having O(n^2) complexity.  Instead I think you could do something like...

```
FieldVector combined_fields = left.output_schema.fields();
const FieldVector& right_fields = right.output_schema.fields();
combined_fields.insert(combined_fields.end(), right_fields.begin(), right_fields.end());
std::shared_ptr<Schema> join_schema = schema(std::move(combined_fields));
```

Reply:
Thank you for catching this one.